dataplane/src/pipeline/static_nf.rs

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use net::buffer::PacketBufferMut;
+use std::marker::PhantomData;
+
+use crate::packet::Packet;
+
+/// Trait for an object that processes a stream of packets.
+///
+/// The `process` method takes an iterator of [`crate::packet::Packet`] objects,
+/// applies the appropriate transformations (or drops) and returns an iterator of
+/// modified packets.
+///
+/// Note that a concrete iterator type is required to call this function and
+/// a concrete iterator type must be returned from this function (i.e., `impl Iterator`).
+pub trait NetworkFunction<Buf: PacketBufferMut> {
+    fn process<'a, Input: Iterator<Item = Packet<Buf>> + 'a>(
+        &'a mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> + 'a;
+}
+
+struct StaticChainImpl<Buf: PacketBufferMut, NF1: NetworkFunction<Buf>, NF2: NetworkFunction<Buf>> {
+    nf1: NF1,
+    nf2: NF2,
+    _marker: PhantomData<Buf>,
+}
+
+impl<Buf: PacketBufferMut, NF1: NetworkFunction<Buf>, NF2: NetworkFunction<Buf>>
+    NetworkFunction<Buf> for StaticChainImpl<Buf, NF1, NF2>
+{
+    fn process<'a, Input: Iterator<Item = Packet<Buf>> + 'a>(
+        &'a mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> + 'a {
+        self.nf2.process(self.nf1.process(input))
+    }
+}
+
+/// Statically chains two [`NetworkFunction`] objects together.
+///
+/// The `chain` method takes two [`NetworkFunction`] objects and returns a new [`NetworkFunction`]
+/// that applies the first function, then the second.
+///
+/// This trait is automatically implemented for all objects that implement [`NetworkFunction`].
+///
+/// > [!WARNING]
+/// > Do not use long chains of statically chained network functions.
```

Comment:
The behavior I saw was more of the "find bugs in LLVM's symbol name length bounds."

I think the warning is good in spirit, but perhaps the details are not ideal.
Even if we explain the full issue, the situation may change and rot the warning.

Perhaps just leave it as "compiler bugs"?  Maybe that is worse.  Not sure :/

Reply:
Also, we might define "long"

Reply:
Done.  I chose 8 as long since that gave ok compile times and didn't OOM the linker.